lib/recursive-methods.rb

```diff
@@ -1,49 +1,98 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - where n is equal to the input value
+# Space complexity: O(n) - where n is equal to the input value
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError, "input must be a number greater than or equal to 0."
+  elsif n == 1 || n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - where n is equal to the length of the input string
```

Comment:
Because `s[0..-2]` will also create a new array, this makes the time complexity also O(n^2)